File: tui/style.ml
include C.Types.Style

module Color = C.Types.Color
module ColorOpt = C.Types.ColorOpt

module Mod = struct
  type t = int

  open Ctypes

  let bold = !@C.Fn.mod_bold |> Unsigned.UInt16.to_int
  let dim = !@C.Fn.mod_dim |> Unsigned.UInt16.to_int
  let italic = !@C.Fn.mod_italic |> Unsigned.UInt16.to_int
  let underlined = !@C.Fn.mod_underlined |> Unsigned.UInt16.to_int
  let slow_blink = !@C.Fn.mod_slow_blink |> Unsigned.UInt16.to_int
  let rapid_blink = !@C.Fn.mod_rapid_blink |> Unsigned.UInt16.to_int
  let reversed = !@C.Fn.mod_reversed |> Unsigned.UInt16.to_int
  let hidden = !@C.Fn.mod_hidden |> Unsigned.UInt16.to_int
  let crossed_out = !@C.Fn.mod_crossed_out |> Unsigned.UInt16.to_int

  let empty = !@C.Fn.mod_empty |> Unsigned.UInt16.to_int
end

let make ?fg ?bg ?(mods = Mod.empty) ?(sub_mods = Mod.empty) () =
  {
    fg =
      (match fg with
      | None -> ColorOpt.None
      | Some color -> ColorOpt.Some color);
    bg =
      (match bg with
      | None -> ColorOpt.None
      | Some color -> ColorOpt.Some color);
    add_modifier = mods;
    sub_modifier = sub_mods;
  }

File: tui/style.rs
use tui::style::Modifier;

#[no_mangle]
pub static tui_mod_bold: u16 = Modifier::BOLD.bits();

#[no_mangle]
pub static tui_mod_dim: u16 = Modifier::DIM.bits();

#[no_mangle]
pub static tui_mod_italic: u16 = Modifier::ITALIC.bits();

#[no_mangle]
pub static tui_mod_underlined: u16 = Modifier::UNDERLINED.bits();

#[no_mangle]
pub static tui_mod_slow_blink: u16 = Modifier::SLOW_BLINK.bits();

#[no_mangle]
pub static tui_mod_rapid_blink: u16 = Modifier::RAPID_BLINK.bits();

#[no_mangle]
pub static tui_mod_reversed: u16 = Modifier::REVERSED.bits();

#[no_mangle]
pub static tui_mod_hidden: u16 = Modifier::HIDDEN.bits();

#[no_mangle]
pub static tui_mod_crossed_out: u16 = Modifier::CROSSED_OUT.bits();

#[no_mangle]
pub static tui_mod_empty: u16 = Modifier::empty().bits();
